File: uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/steps/ListHabitsSteps.java
/*
 * Copyright (C) 2017 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.acceptance.steps;

import android.support.test.espresso.*;
import android.view.*;

import org.hamcrest.*;
import org.isoron.uhabits.R;
import org.isoron.uhabits.activities.habits.list.views.*;

import java.util.*;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.matcher.ViewMatchers.hasDescendant;
import static android.support.test.espresso.matcher.ViewMatchers.isEnabled;
import static android.support.test.espresso.matcher.ViewMatchers.withClassName;
import static android.support.test.espresso.matcher.ViewMatchers.withContentDescription;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withParent;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.CoreMatchers.*;
import static org.isoron.uhabits.BaseUserInterfaceTest.device;
import static org.isoron.uhabits.acceptance.steps.CommonSteps.clickText;

public abstract class ListHabitsSteps
{
    public static void clickMenu(MenuItem item)
    {
        switch (item)
        {
            case ABOUT:
                clickTextInsideOverflowMenu(R.string.about);
                break;

            case HELP:
                clickTextInsideOverflowMenu(R.string.help);
                break;

            case SETTINGS:
                clickTextInsideOverflowMenu(R.string.settings);
                break;

            case ADD:
                clickViewWithId(R.id.actionAdd);
                break;

            case EDIT:
                clickViewWithId(R.id.action_edit_habit);
                break;

            case DELETE:
                clickTextInsideOverflowMenu(R.string.delete);
                break;

            case ARCHIVE:
                clickTextInsideOverflowMenu(R.string.archive);
                break;

            case UNARCHIVE:
                clickTextInsideOverflowMenu(R.string.unarchive);
                break;

            case TOGGLE_ARCHIVED:
                clickViewWithId(R.id.action_filter);
                clickText(R.string.hide_archived);
                break;

            case TOGGLE_COMPLETED:
                clickViewWithId(R.id.action_filter);
                clickText(R.string.hide_completed);
                break;
        }
    }

    private static void clickTextInsideOverflowMenu(int id)
    {
        onView(allOf(withContentDescription("More options"), withParent(
            withParent(withClassName(endsWith("Toolbar")))))).perform(click());

        onView(withText(id)).perform(click());
    }

    private static void clickViewWithId(int id)
    {
        onView(withId(id)).perform(click());
    }

    private static ViewAction longClickDescendantWithClass(Class cls, int count)
    {
        return new ViewAction()
        {

            @Override
            public Matcher<View> getConstraints()
            {
                return isEnabled();
            }

            @Override
            public String getDescription()
            {
                return "perform on children";
            }

            @Override
            public void perform(UiController uiController, View view)
            {
                LinkedList<ViewGroup> stack = new LinkedList<>();
                if (view instanceof ViewGroup) stack.push((ViewGroup) view);
                int countRemaining = count;

                while (!stack.isEmpty())
                {
                    ViewGroup vg = stack.pop();
                    for (int i = 0; i < vg.getChildCount(); i++)
                    {
                        View v = vg.getChildAt(i);
                        if (v instanceof ViewGroup) stack.push((ViewGroup) v);
                        if (cls.isInstance(v) && countRemaining > 0)
                        {
                            v.performLongClick();
                            uiController.loopMainThreadUntilIdle();
                            countRemaining--;
                        }
                    }
                }
            }
        };
    }

    public static void longPressCheckmarks(String habit, int count)
    {
        CommonSteps.scrollToText(habit);
        onView(allOf(hasDescendant(withText(habit)),
            withClassName(endsWith("HabitCardView")))).perform(
            longClickDescendantWithClass(CheckmarkButtonView.class, count));
        device.waitForIdle();
    }

    public enum MenuItem
    {
        ABOUT, HELP, SETTINGS, EDIT, DELETE, ARCHIVE, TOGGLE_ARCHIVED,
        UNARCHIVE, TOGGLE_COMPLETED, ADD
    }
}

File: uhabits-android/src/androidTest/java/org/isoron/uhabits/acceptance/steps/ListHabitsSteps.kt
/*
 * Copyright (C) 2016-2021 Álinson Santos Xavier <git@axavier.org>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.isoron.uhabits.acceptance.steps

import android.view.View
import android.view.ViewGroup
import androidx.test.espresso.Espresso
import androidx.test.espresso.UiController
import androidx.test.espresso.ViewAction
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.matcher.ViewMatchers
import org.hamcrest.CoreMatchers.allOf
import org.hamcrest.CoreMatchers.endsWith
import org.hamcrest.Matcher
import org.isoron.uhabits.BaseUserInterfaceTest
import org.isoron.uhabits.BaseUserInterfaceTest.Companion.device
import org.isoron.uhabits.R
import org.isoron.uhabits.activities.habits.list.views.CheckmarkButtonView
import java.util.LinkedList

object ListHabitsSteps {
    fun clickMenu(item: MenuItem?) {
        when (item) {
            MenuItem.ABOUT -> clickTextInsideOverflowMenu(R.string.about)
            MenuItem.HELP -> clickTextInsideOverflowMenu(R.string.help)
            MenuItem.SETTINGS -> clickTextInsideOverflowMenu(R.string.settings)
            MenuItem.ADD -> clickViewWithId(R.id.actionCreateHabit)
            MenuItem.EDIT -> clickViewWithId(R.id.action_edit_habit)
            MenuItem.DELETE -> clickTextInsideOverflowMenu(R.string.delete)
            MenuItem.ARCHIVE -> clickTextInsideOverflowMenu(R.string.archive)
            MenuItem.UNARCHIVE -> clickTextInsideOverflowMenu(R.string.unarchive)
            MenuItem.TOGGLE_ARCHIVED -> {
                clickViewWithId(R.id.action_filter)
                CommonSteps.clickText(R.string.hide_archived)
            }
            MenuItem.TOGGLE_COMPLETED -> {
                clickViewWithId(R.id.action_filter)
                CommonSteps.clickText(R.string.hide_completed)
            }
        }
        device.waitForIdle()
    }

    private fun clickTextInsideOverflowMenu(id: Int) {
        Espresso.onView(
            allOf(
                ViewMatchers.withContentDescription("More options"),
                ViewMatchers.withParent(
                    ViewMatchers.withParent(
                        ViewMatchers.withClassName(
                            endsWith("Toolbar")
                        )
                    )
                )
            )
        ).perform(ViewActions.click())
        Espresso.onView(ViewMatchers.withText(id)).perform(ViewActions.click())
    }

    private fun clickViewWithId(id: Int) {
        Espresso.onView(ViewMatchers.withId(id)).perform(ViewActions.click())
    }

    private fun longClickDescendantWithClass(cls: Class<*>, count: Int): ViewAction {
        return object : ViewAction {
            override fun getConstraints(): Matcher<View> {
                return ViewMatchers.isEnabled()
            }

            override fun getDescription(): String {
                return "perform on children"
            }

            override fun perform(uiController: UiController, view: View) {
                val stack = LinkedList<ViewGroup>()
                if (view is ViewGroup) stack.push(view)
                var countRemaining = count
                while (!stack.isEmpty()) {
                    val vg = stack.pop()
                    for (i in 0 until vg.childCount) {
                        val v = vg.getChildAt(i)
                        if (v is ViewGroup) stack.push(v)
                        if (cls.isInstance(v) && countRemaining > 0) {
                            v.performLongClick()
                            uiController.loopMainThreadUntilIdle()
                            countRemaining--
                        }
                    }
                }
            }
        }
    }

    fun longPressCheckmarks(habit: String?, count: Int) {
        CommonSteps.scrollToText(habit)
        Espresso.onView(
            allOf(
                ViewMatchers.hasDescendant(ViewMatchers.withText(habit)),
                ViewMatchers.withClassName(endsWith("HabitCardView"))
            )
        ).perform(
            longClickDescendantWithClass(CheckmarkButtonView::class.java, count)
        )
        BaseUserInterfaceTest.device.waitForIdle()
    }

    fun changeSort(sortText: String) {
        clickViewWithId(R.id.action_filter)
        Espresso.onView(ViewMatchers.withText("Sort")).perform(ViewActions.click())
        Espresso.onView(ViewMatchers.withText(sortText)).perform(ViewActions.click())
    }

    enum class MenuItem {
        ABOUT, HELP, SETTINGS, EDIT, DELETE, ARCHIVE, TOGGLE_ARCHIVED, UNARCHIVE, TOGGLE_COMPLETED, ADD
    }
}
